### PR TITLE
MOM6: +Corrections to ice shelf-related code (includes PR#1278)

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
This PR adds runtime parameters that recover the answers to the ISOMIP test
cases that were changed without acknowledgement or documentation as a (small)
part of MOM6 PR#1265, and it adds in two new ice_shelf-related initialization
subroutines that remove the need to call initialize_ice_shelf multiple times.
The changes in MOM6 PR#1278 had not yet been merged into the main branch because
the full regression testing is not working, so that PR has been subsumed into
this larger PR.  There are new entries in many MOM_parameter_doc.all files as
a result of this PR, and in cases with the ice shelves enabled all parameters
from the MOM_input files are once again being logged.  All answers in the
MOM6-examples test suite, including those of the expensive and infrequently run
test cases, are once again back to their values prior to MOM6 PR#1265.

  The self-consistency testing cases tc1-tc3 all pass, but I am still unable to
run tc4 due to python issues, and tc4 might show up as a regression failure due
to recent changes to activate the ice shelf component and the corrections to
recover prior answers in this PR.

  The commits in this PR (including those that were in MOM6 PR#1278) include:

- NOAA-GFDL/MOM6@0ddfd8725 Merge branch 'dev/gfdl' into ice_shelf_fix
- NOAA-GFDL/MOM6@a1167461b +Added initialize_ice_shelf_fluxes
(The following commits are a part of MOM6 PR#1278):
- NOAA-GFDL/MOM6@aa26c6005 Merge branch 'dev/gfdl' into revert_drivers
- NOAA-GFDL/MOM6@d2d047c5e remove masking from ice shelf diagnostics
- NOAA-GFDL/MOM6@88c4102d6 move complete_initialization return before diag chksums
- NOAA-GFDL/MOM6@44e80d59f remove rotation related changes in initialize_ice_shelf
- NOAA-GFDL/MOM6@bdcb8588a Move call to end ice shelf diag mediator into ice_shelf_end
- NOAA-GFDL/MOM6@19c1a6af2 fix compile issues
- NOAA-GFDL/MOM6@55c6eaa57 Modifications for coupled driver
- NOAA-GFDL/MOM6@8087f9300 Return early from ice shelf initialization to avoid registering diagnostics and restarts during the call from initialize_MOM
- NOAA-GFDL/MOM6@f3dc73f7b Revert forcing,fluxes and sfc_state from pointers.
- NOAA-GFDL/MOM6@2c93933a9 Revert changes in the MOM drivers for ice shelf initialization

### CODE REVIEW

File: src/ice_shelf/MOM_ice_shelf.F90 [hunk 1]
```diff
@@ -1751,33 +1679,73 @@ subroutine initialize_ice_shelf(param_file, ocn_grid, Time, CS, diag, forces_in,
...
+       u_desc = var_desc("taux_shelf", "Pa", "the zonal stress on the ocean under ice shelves", &
+            hor_grid='Cu',z_grid='1')
+       v_desc = var_desc("tauy_shelf", "Pa", "the meridional stress on the ocean under ice shelves", &
+            hor_grid='Cv',z_grid='1')
+       call register_restart_pair(sfc_state%taux_shelf, sfc_state%tauy_shelf, u_desc,v_desc, &
+            .false., CS%restart_CSp)
+    endif
+  endif
+
+  call register_restart_field(ISS%h_shelf, "_shelf", .true., CS%restart_CSp, &
```
Comment (gustavo-marques) [inline src/ice_shelf/MOM_ice_shelf.F90:1714]: This field is being registered twice. It was already registered above as "h_shelf".

Comment (Hallberg-NOAA) [reply src/ice_shelf/MOM_ice_shelf.F90]: Thank you very much for catching this duplicate call. This PR has been updated to remove the duplicate call.

  We had a significant breach in our testing protocols when we accepted PR #1265, which allowed this problem and the others that are addressed by this PR to slip into the dev/gfdl version of the MOM6 code.  The problems with that PR included widespread structural changes mixed in with answer changes with no run-time option to recover previous answers, limited discussion and no group consensus on significant restructuring, forced changes to top-level interfaces without adequate coordination across the various institutional forks, a lack of enforcement of MOM6 code style standards, and enormous commits within the PR that did not facilitate debugging. I hope that we have now corrected most of the fallout from PR #1265, but it has clearly exposed weaknesses in our testing and evaluation procedures that we will need to correct.

**Review by MJHarrison-GFDL — COMMENTED** [review]
@Hallberg-NOAA  I agree with these changes and thank you for separating the forcing initialization which is a much simpler and cleaner solution.

**Comment by adcroft** [comment]
https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/11786

**Comment by adcroft** [comment]
Tests passed but doc files were changed so I'll handle this manually

**Review by adcroft — APPROVED** [review]

**Pull request merged**